velox/exec/fuzzer/CacheFuzzer.cpp

```diff
@@ -0,0 +1,358 @@
+
+  // TODO: Test different ssd write behaviors with AsyncDataCache::Options.
+  cache_ = AsyncDataCache::create(allocator_, std::move(ssdCache), {});
+}
+
+void CacheFuzzer::initializeInputs() {
+  auto readOptions = io::ReaderOptions(pool_.get());
+  auto tracker = std::make_shared<ScanTracker>(
+      "testTracker", nullptr, io::ReaderOptions::kDefaultLoadQuantum);
+  auto ioStats = std::make_shared<IoStatistics>();
+  inputs_.reserve(FLAGS_num_source_files);
+  for (auto i = 0; i < FLAGS_num_source_files; ++i) {
+    // Initialize buffered input.
+    auto readFile = fs_->openFileForRead(fileNames_[i]);
+    groupIds_.emplace_back(
+        fileIds(), fmt::format("group{}", i / FLAGS_num_files_in_group));
+    auto const withExecutor =
+        boost::random::uniform_int_distribution<int8_t>(0, 1)(rng_) == 1;
+    inputs_.emplace_back(std::make_unique<CachedBufferedInput>(
+        std::move(readFile),
+        MetricsLog::voidLog(),
+        fileIds_[i].id(), // NOLINT
+        cache_.get(),
+        tracker,
+        groupIds_[i].id(),
+        ioStats,
+        withExecutor ? executor_.get() : nullptr,
+        readOptions));
+
+    // Divide file into fragments.
+    std::vector<std::pair<int32_t, int32_t>> fragments;
+    int32_t offset = 0;
+    while (offset < fileSizes_[i]) {
+      const auto length = boost::random::uniform_int_distribution<int32_t>(
+          1, fileSizes_[i] - offset)(rng_);
+      fragments.emplace_back(offset, length);
+      offset += length;
+    }
+    fileFragments_.emplace_back(std::move(fragments));
+  }
+}
+
+void CacheFuzzer::readCache() {
+  runThreads(FLAGS_num_threads, [&](int32_t /*i*/) {
+    auto numReads = boost::random::uniform_int_distribution<int32_t>(
+        10, FLAGS_max_num_reads)(rng_);
+    while (numReads-- > 0) {
+      const auto fileIdx = boost::random::uniform_int_distribution<int32_t>(
+          0, FLAGS_num_source_files - 1)(rng_);
+      const auto fragmentIdx = boost::random::uniform_int_distribution<int32_t>(
+          0, fileFragments_[fileIdx].size() - 1)(rng_);
+      read(fileIdx, fragmentIdx);
+    }
+  });
+}
+
+void CacheFuzzer::reset() {
+  cache_->shutdown();
+  cache_->ssdCache()->testingWaitForWriteToFinish();
+  executor_->join();
+  executor_.reset();
+  fileNames_.clear();
+  fileIds_.clear();
+  fileSizes_.clear();
+  fileFragments_.clear();
+  groupIds_.clear();
+  inputs_.clear();
+  fs_.reset();
+  cache_.reset();
+  memoryManager_.reset();
+  sourceDataDir_.reset();
+  fileIds().testingReset();
+}
+
+uint64_t CacheFuzzer::read(uint32_t fileIdx, int32_t fragmentIdx) {
+  // TODO: Faulty injection.
+  const auto offset = fileFragments_[fileIdx][fragmentIdx].first;
+  const auto length = fileFragments_[fileIdx][fragmentIdx].second;
+  auto stream = inputs_[fileIdx]->read(offset, length, LogType::TEST);
+  const void* buffer;
+  int32_t size;
+  int32_t numRead = 0;
+  while (numRead < length) {
+    if (stream->Next(&buffer, &size)) {
+      numRead += size;
```

Comment:
Shall we verify the read content here?

Reply:
I was planning to have it added in follow-up.

Reply:
Please add in this PR and I think it is trivial based on the file offset? Thanks!